brave/brave-core
# Tor profile rework

Fix https://github.com/brave/brave-browser/issues/2307

It would be easier to review this PR per commit.
What this PR does:
- Tor profile is not based on guest profiles anymore, when SwitchToTorProfile is called, we will create a dummy regular Tor profile and use its off-the-record profile to browse. 
-  Service factories are hooked to inherit bookmarks(3418ed78a120fac589edc0bf6be8c94befbc6208), preferences(0979ed1c1c0ae3524d5bf5869fdb9c3f8338d85c) from its parent profile (the one that is used before switching to Tor profile). Our behavior should be the same with private window for these two.
- Navigation requests to pages like brave://settings will be redirected back to its parent profile. 
(601b6f1f69e68795bf41a0c3afd6643286520d06)
- unit and browser tests

## Submitter Checklist:
- Verified that these changes build without errors on
- Verified that these changes pass automated tests (unit, browser, security tests) on

## Test Plan:
1. Open the browser and go to brave://settings and disable `Show bookmarks bar`.
2. Visit https://check.torproject.org and add the page into bookmarks.
3. Open a tor window, the new tab page bookmark bar should show the one we added in step 2.
4. Visit https://check.torproject.org in Tor window, it should show we're using Tor network.
5. Go back to the normal window and go to brave://settings and enable `Show bookmarks bar`
6. Go back to the https://check.torproject.org in Tor window, it should now display the bookmark bar with the bookmark we added in step 2.
7. Add a bookmark again in the normal window, it should show up in the Tor window too.
8. Delete a bookmark from the Tor window, it should disappear in the normal window too.
9. Delete a bookmark from the normal window, it should disappear in the Tor window too.

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
